modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferBase.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+namespace Drupal\commerce_promotion\Plugin\Commerce\PromotionOffer;
+
+use Drupal\commerce_order\AdjustableInterface;
+use Drupal\commerce_order\Adjustment;
+use Drupal\commerce_price\Price;
+use Drupal\Core\Executable\ExecutablePluginBase;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Plugin\ContextAwarePluginAssignmentTrait;
+
+/**
+ * Base class for Promotion Offer plugins.
+ */
+abstract class PromotionOfferBase extends ExecutablePluginBase implements PromotionOfferInterface {
+
+  use ContextAwarePluginAssignmentTrait;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+
+    $this->setConfiguration($configuration);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function appliesTo() {
+    return $this->pluginDefinition['target'];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getTargetEntity() {
+    return $this->getContextValue($this->appliesTo());
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function applyAdjustment(AdjustableInterface $item, Price $amount) {
+    $item->addAdjustment(new Adjustment([
+      'type' => 'promotion',
+      'label' => $this->pluginDefinition['label'],
```

Comment:
Let's default it to t('Discount'), that's a safe option other systems (and discount 1.x) use, then and add a todo for #2770731.